[Xtensa] Implement Xtensa Floating Point Option.

andreisfr opened the pull request.

andreisfr committed:
[Xtensa] Implement Xtensa Floating Point Option.

arsenm commented on llvm/lib/Target/Xtensa/Disassembler/XtensaDisassembler.cpp:
```diff
@@ -113,6 +113,44 @@ static DecodeStatus DecodeMR23RegisterClass(MCInst &Inst, uint64_t RegNo,
   return MCDisassembler::Success;
 }
 
+static const unsigned FPRDecoderTable[] = {
```
```suggestion
static const MCPhysReg FPRDecoderTable[] = {
```

andreisfr replied:
fixed

arsenm commented on llvm/lib/Target/Xtensa/Disassembler/XtensaDisassembler.cpp:
```diff
@@ -113,6 +113,44 @@ static DecodeStatus DecodeMR23RegisterClass(MCInst &Inst, uint64_t RegNo,
+  if (RegNo >= std::size(FPRDecoderTable))
+    return MCDisassembler::Fail;
+
+  unsigned Reg = FPRDecoderTable[RegNo];
```
```suggestion
  MCPhysReg Reg = FPRDecoderTable[RegNo];
```

arsenm commented on llvm/lib/Target/Xtensa/XtensaISelLowering.cpp:
```diff
@@ -175,6 +184,103 @@ XtensaTargetLowering::XtensaTargetLowering(const TargetMachine &TM,
   setOperationAction(ISD::VACOPY, MVT::Other, Custom);
   setOperationAction(ISD::VAEND, MVT::Other, Expand);
 
+  // Handle floating-point types.
```
This is mostly reinventing the default logic based on legal types

arsenm commented on llvm/lib/Target/Xtensa/XtensaISelLowering.cpp:
```diff
@@ -185,6 +291,11 @@ bool XtensaTargetLowering::isOffsetFoldingLegal(
   return false;
 }
 
+bool XtensaTargetLowering::isFPImmLegal(const APFloat &Imm, EVT VT,
+                                        bool ForCodeSize) const {
+  return false;
+}
```
I believe this is redundant if you make ConstantFP Expand

andreisfr replied:
fixed

arsenm commented on llvm/lib/Target/Xtensa/XtensaISelLowering.cpp:
```diff
@@ -815,6 +936,21 @@ SDValue XtensaTargetLowering::LowerImmediate(SDValue Op,
   return Op;
 }
 
+SDValue XtensaTargetLowering::LowerImmediateFP(SDValue Op,
```
This is the same as the default expansion?

arsenm replied:
Can this just use the default expand?

arsenm commented on llvm/test/CodeGen/Xtensa/float-arith.ll:
```diff
@@ -0,0 +1,569 @@
+
+declare float @llvm.fabs.f32(float)
+
+define float @fabs_s(float %a, float %b) nounwind {
```
Missing fneg and copysign tests

arsenm commented on llvm/test/CodeGen/Xtensa/float-arith.ll:
```diff
@@ -0,0 +1,569 @@
+; XTENSA-NEXT:    add.s f8, f9, f8
+; XTENSA-NEXT:    rfr a2, f8
+; XTENSA-NEXT:    ret
+  %1 = fadd float %a, %b
```
Use named values in tests

arsenm commented on llvm/lib/Target/Xtensa/XtensaISelLowering.cpp:
```diff
@@ -71,6 +75,8 @@ XtensaTargetLowering::XtensaTargetLowering(const TargetMachine &TM,
 
   setOperationAction(ISD::Constant, MVT::i32, Custom);
   setOperationAction(ISD::Constant, MVT::i64, Expand);
+  setOperationAction(ISD::ConstantFP, MVT::f32, Custom);
+  setOperationAction(ISD::ConstantFP, MVT::f64, Expand);
```
The custom looks the same as default expand

andreisfr committed:
[Xtensa] Fix Floating Point implementation.

Add more more FP tests. Remove redundant code.
Improve User Registers parsing.

arsenm commented on llvm/lib/Target/Xtensa/Disassembler/XtensaDisassembler.cpp:
```diff
@@ -113,6 +113,39 @@ static DecodeStatus DecodeMR23RegisterClass(MCInst &Inst, uint64_t RegNo,
+static DecodeStatus DecodeURRegisterClass(MCInst &Inst, uint64_t RegNo,
+                                          uint64_t Address,
+                                          const MCDisassembler *Decoder) {
+  const llvm::MCSubtargetInfo STI =
+      ((const MCDisassembler *)Decoder)->getSubtargetInfo();
+
```
```suggestion
```

Don't make a copy of SubtargetInfo, but this also isn't used since it's requeryed below anyway

andreisfr replied:
Thank you very much for comment. I fixes this.

andreisfr replied:
@arsenm , I'm sorry, may I ask you also for help with review of this PR https://github.com/llvm/llvm-project/pull/137135?

andreisfr replied:
fixed

andreisfr committed:
[Xtensa] Minor fixes.

arsenm commented on llvm/lib/Target/Xtensa/XtensaISelLowering.cpp:
```diff
@@ -815,6 +898,21 @@ SDValue XtensaTargetLowering::LowerImmediate(SDValue Op,
+  const ConstantFPSDNode *CN = cast<ConstantFPSDNode>(Op);
+  SDLoc DL(CN);
+  APFloat apval = CN->getValueAPF();
+  int64_t value = llvm::bit_cast<uint32_t>(CN->getValueAPF().convertToFloat());
```
Keep this in APFloat instead of using convertToFloat (i.e. bitcastToAPInt)

andreisfr replied:
Fixed.

arsenm commented on llvm/lib/Target/Xtensa/XtensaISelLowering.cpp:
```diff
@@ -815,6 +898,21 @@ SDValue XtensaTargetLowering::LowerImmediate(SDValue Op,
+  if (Op.getValueType() == MVT::f32) {
+    Type *Ty = Type::getInt32Ty(*DAG.getContext());
+    Constant *CV = ConstantInt::get(Ty, value);
+    SDValue CP = DAG.getConstantPool(CV, MVT::i32);
+    return DAG.getNode(ISD::BITCAST, DL, MVT::f32, CP);
```
This usage of ConstantPool looks broken, it gives a pointer that you're supposed to load the value from. How can you interpret the pointer value as an FP type?

andreisfr replied:
I think you are right, thanks for the comment. I have significantly reworked the code related to loading constants from the constant pool.  So, in last version FP immediate is just expanded instead of custom lowering. Also added changes that fix a possible memory leak when creating XtensaConstantPoolValue objects.

andreisfr replied:
fixed

andreisfr committed:
[Xtensa] Fix ConstantPool lowering.

Fix lowering FP immediate constant. Fix loading from constant pool callee, basic block,
globaladdress and jumptable addresses. Also fixed potential memory leakage when
several similar XtensaConstantPoolValue objects are created, implement unique ID for such objects.

andreisfr committed:
[Xtensa] Implement pattern ordering for load from ConstnatPool.

andreisfr committed:
[Xtensa] Minor fix in XtensaAsnParser. Remove redundant code.

andreisfr commented on llvm/lib/Target/Xtensa/AsmParser/XtensaAsmParser.cpp:
```diff
@@ -596,25 +597,33 @@ ParseStatus XtensaAsmParser::parseRegister(OperandVector &Operands,
     if (ReadCount == 2 && Buf[1].getKind() == AsmToken::RParen) {
-      if ((Buf[0].getKind() == AsmToken::Integer) && (!SR))
+      if ((Buf[0].getKind() == AsmToken::Integer) &&
+          (RegType == Xtensa_Generic))
```
Fixed

MaskRay commented on llvm/lib/Target/Xtensa/Disassembler/XtensaDisassembler.cpp:
```diff
@@ -113,6 +113,36 @@ static DecodeStatus DecodeMR23RegisterClass(MCInst &Inst, uint64_t RegNo,
   return MCDisassembler::Success;
 }
 
+static const MCPhysReg FPRDecoderTable[] = {
```
namespace-scope const implies internal linkage; can omit `static`

andreisfr replied:
Fixed.

MaskRay commented on llvm/lib/Target/Xtensa/Disassembler/XtensaDisassembler.cpp:
```diff
@@ -113,6 +113,36 @@ static DecodeStatus DecodeMR23RegisterClass(MCInst &Inst, uint64_t RegNo,
 }
 
+static const MCPhysReg FPRDecoderTable[] = {
+    Xtensa::F0,  Xtensa::F1,  Xtensa::F2,  Xtensa::F3, Xtensa::F4,  Xtensa::F5,
```
Can assume the values are consecutive. See https://reviews.llvm.org/D67423 (TableGen `LessRecordRegister`)

andreisfr committed:
[Xtensa] Minor improvements in disassembler.

MaskRay commented on llvm/lib/Target/Xtensa/MCTargetDesc/XtensaMCTargetDesc.cpp:
```diff
@@ -111,6 +114,17 @@ bool Xtensa::checkRegister(MCRegister RegNo, const FeatureBitset &FeatureBits) {
+// Get Xtensa User Register by encoding value.
+MCRegister Xtensa::getUserRegister(unsigned Code) {
+  switch (Code) {
+  case 232:
```
Where does the magic value come from?

andreisfr replied:
I modified the function to avoid using register codes.

andreisfr committed:
[Xtensa] Minor fix of the Xtensa::getUserRegister function.

andreisfr committed:
[Xtensa] Fix lowering immediate.

andreisfr commented:
I rebased PR on latest upstream version. Also implemented fix  of the immediate lowering.

andreisfr committed:
[Xtensa] Minor fix.
